File: docs/src/main/paradox/tutorial/networking/virtual/pingpong-virtual/src/main/java/se/sics/test/TAddressConverter.java
package se.sics.test;

import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.Map;
import se.sics.kompics.config.Conversions;
import se.sics.kompics.config.Converter;

public class TAddressConverter implements Converter<TAddress> {

    @Override
    public TAddress convert(Object o) {
        if (o instanceof Map) {
            try {
                Map m = (Map) o;
                String hostname = Conversions.convert(m.get("host"), String.class);
                int port = Conversions.convert(m.get("port"), Integer.class);
                InetAddress ip = InetAddress.getByName(hostname);
                return new TAddress(ip, port);
            } catch (UnknownHostException ex) {
                return null;
            }
        }
        if (o instanceof String) {
            try {
                String[] ipport = ((String) o).split(":");
                InetAddress ip = InetAddress.getByName(ipport[0]);
                int port = Integer.parseInt(ipport[1]);
            } catch (UnknownHostException ex) {
                return null;
            }
        }
        return null;
    }

    @Override
    public Class<TAddress> type() {
        return TAddress.class;
    }

}

File: docs/src/main/scala/sexamples/simulation/pingpong/TAddressConverter.scala
package sexamples.simulation.pingpong

import java.net.{InetAddress, UnknownHostException}
import java.util.Map
import se.sics.kompics.config.{Conversions, Converter}

object TAddressConverter extends Converter[TAddress] {

  override def convert(o: AnyRef): TAddress = {
    o match {
      case m: Map[String, Any] @unchecked => {
        val hostname = Conversions.convert(m.get("host"), classOf[String]);
        val port: Int = Conversions.convert(m.get("port"), classOf[Integer]);
        val ip = InetAddress.getByName(hostname);
        TAddress(ip, port)
      }
      case s: String => {
        val ipport = s.split(":");
        val ip = InetAddress.getByName(ipport(0));
        val port = Integer.parseInt(ipport(1));
        TAddress(ip, port)
      }
      case _ => null
    }
  }

  override def `type`(): Class[TAddress] = classOf[TAddress];
}
